python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -222,54 +225,62 @@ def _update(self):
             # of that type is lower/equal to the min_workers of that type
             # or it is needed for request_resources().
-            should_keep_worker_of_node_type, node_type_counts = \
-                self._keep_worker_of_node_type(node_id, node_type_counts)
-            if ((should_keep_worker_of_node_type
+            should_keep_or_terminate = self._keep_worker_of_node_type(
+                node_id, node_type_counts)
+            if ((should_keep_or_terminate == KeepOrTerminate.keep
                  or not nodes_allowed_to_terminate.get(node_id, True))
                     and self.launch_config_ok(node_id)):
+                keep_node(node_id)
+                continue
+            if should_keep_or_terminate == KeepOrTerminate.terminate:
+                terminate_node(node_id, "max_workers_per_type")
                 continue
 
             node_ip = self.provider.internal_ip(node_id)
             if node_ip in last_used and last_used[node_ip] < horizon:
-                logger.info("StandardAutoscaler: "
-                            "{}: Terminating idle node.".format(node_id))
-                self.event_summarizer.add(
-                    "Removing {} nodes of type " + self._get_node_type(node_id)
-                    + " (idle).",
-                    quantity=1,
-                    aggregate=operator.add)
-                nodes_to_terminate.append(node_id)
+                terminate_node(node_id, "idle")
             elif not self.launch_config_ok(node_id):
-                logger.info("StandardAutoscaler: "
-                            "{}: Terminating outdated node.".format(node_id))
-                self.event_summarizer.add(
-                    "Removing {} nodes of type " + self._get_node_type(node_id)
-                    + " (outdated).",
-                    quantity=1,
-                    aggregate=operator.add)
-                nodes_to_terminate.append(node_id)
-
-        if nodes_to_terminate:
-            self._terminate_nodes_and_cleanup(nodes_to_terminate)
-            nodes = self.workers()
-
+                terminate_node(node_id, "outdated")
+            else:
+                keep_node(node_id)
+                nodes_we_could_terminate.append(node_id)
         # Terminate nodes if there are too many
-        nodes_to_terminate = []
-        while (len(nodes) -
-               len(nodes_to_terminate)) > self.config["max_workers"] and nodes:
-            to_terminate = nodes.pop()
-            logger.info("StandardAutoscaler: "
-                        "{}: Terminating unneeded node.".format(to_terminate))
-            self.event_summarizer.add(
-                "Removing {} nodes of type " +
-                self._get_node_type(to_terminate) + " (max workers).",
-                quantity=1,
-                aggregate=operator.add)
-            nodes_to_terminate.append(to_terminate)
+        num_nodes_to_terminate = len(nodes_to_keep) - \
```

Comment:
this should be `len(nodes) - self.config["max_workers"]` what am I missing?

Reply:
nodes : self.workers() the the beginning of the _update.
nodes_to_keep are nodes that are not scheduled for termination (due to max_workers per type).